### DISCUSSION TITLE
Now with maps and structs!

### DISCUSSION THREAD

**Discussion by burakemir** [google/mangle#11] [Announcements]
And our first announcement: Mangle has maps and structs. These are useful for representing structured data, similar to JSON or protocol buffers.

Proper documentation is in the works, but in short, there are built-in functions `fn:map` and `fn:struct` which construct maps and struct constants. Until then, you may want to check out some unit tests.

A map expression is simply an apply expression `fn:map( key_1, val_1, ..., key_n, val_n)` with a convenient syntax `[ key_1 : val_1, ..., key_n : val_n ]`. A type expression for map types is `fn:Map(Key, Val)` and it conforms to .`fn:Map(Key2, Val2)` if `Key2` conforms to `Key1` and `Val1` conforms to `Val2`.

Structs are written `fn:struct( key_1, val_1, ..., key_n, val_n)` with a convenient syntax `{ key_1 : val_1, ..., key_n : val_n }` A type expression for struct types is `fn:Struct(key_1, Val_1, ..., key_n, Val_n)` where each `key_i` is a name constant, and it conforms to another struct type if it is "longer" and has conforming values for matching keys.

In both cases, the convenient syntax is shorthand for the apply expression (it still involves a function symbol). Any rule that involves apply expressions could construct an unbounded number of tuples (unlike a proper datalog rule), and one has to be careful about non-termination. 

Accessing data from maps and structs happens with the new built-in match predicates `:match_entry(Map, Entry, Value)` (for maps) and `:match_field(Struct, Key, Value)` or by checking equality.

**Comment by tdunning** [comment]
This comment is pretty late relative to your announcement, but a key question relative to maps is whether there is an aggregator that collects keys and values into a map?

And can maps or lists be used as a fact base? Would `:match_entry(Map, Entry, Value)` work for that if `Entry` and `Value` are unbound?

**Reply by burakemir** [reply]
For aggregation into map: this does not exist. Great feature request! Filed #72 

About :match_entry(Map, Entry, Value): you have to provide Map and Entry, and Value will be bound (= is an output parameter).

**Comment by tdunning** [comment]
Thanks!
